src/app/file/pixly_format.cpp

```diff
@@ -264,37 +153,72 @@ bool PixlyFormat::onLoad(FileOp* fop)
       visible[(int)layer_index] += (int)(std::string(check(xmlFrame->Attribute("visible"),"false")) == "true");
 
       int x0 = check_number<int>(xmlRegion->Attribute("x"));
-      int y0 = check_number<int>(xmlRegion->Attribute("y")); // inverted
+      int y0_up = check_number<int>(xmlRegion->Attribute("y")); // inverted
 
-      if(y0 < 0 || y0 + frameHeight > imageHeight || x0 < 0 || x0 + frameWidth > imageWidth) {
+      if(y0_up < 0 || y0_up + frameHeight > sheetHeight || x0 < 0 || x0 + frameWidth > sheetWidth) {
         throw Exception("looking for cels outside the bounds of the PNG");
       }
 
-      base::UniquePtr<Cel> cel;
-      ImageRef image(Image::create(pixelFormat, frameWidth, frameHeight));
+      // read cel images
+      ImageRef image(Image::create(IMAGE_RGB, frameWidth, frameHeight));
 
-      // Convert rows_pointer into the doc::Image
       for (int y = 0; y < frameHeight; y++) {
         // RGB_ALPHA
-        uint8_t* src_address = rows_pointer[imageHeight-1 - y0 - (frameHeight-1) + y] + (x0 * 4);
-        uint32_t* dst_address = (uint32_t*)image->getPixelAddress(0, y);
-        unsigned int r, g, b, a;
-
-        for (int x=0; x<frameWidth; x++) {
-          r = *(src_address++);
-          g = *(src_address++);
-          b = *(src_address++);
-          a = *(src_address++);
-          *(dst_address++) = rgba(r, g, b, a);
-        }
+        int y0_down = sheetHeight-1 - y0_up - (frameHeight-1) + y;
+        uint32_t* src_begin = (uint32_t*)sheet->getPixelAddress(x0           , y0_down);
+        uint32_t* src_end   = (uint32_t*)sheet->getPixelAddress(x0+frameWidth, y0_down);
+        uint32_t* dst_begin = (uint32_t*)image->getPixelAddress(0, y);
+
+        std::copy(src_begin, src_end, dst_begin);
       }
 
-      cel.reset(new Cel(frame, image));
-      static_cast<LayerImage*>(layer)->addCel(cel);
-      cel.release();
+      // make cel trimmed or empty
+      gfx::Rect bounds;
+      if (algorithm::shrink_bounds(image.get(), bounds, image->maskColor())) {
+        ImageRef trim_image(crop_image(image.get(),
+                                  bounds.x, bounds.y,
+                                  bounds.w, bounds.h,
+                                  image->maskColor()));
+
+
+        Cel* cel = NULL;
+        if((int)frame > 0) {
+          // link identical neighbors
+          Cel *prev_cel = static_cast<LayerImage*>(layer)->cel(frame-1);
+          if(prev_cel && prev_cel->x() == bounds.x && prev_cel->y() == bounds.y) {
+            Image *prev_image = prev_cel->image();
+            if(prev_image && prev_image->width() == bounds.w && prev_image->height() == bounds.h) {
```

Comment:
Replace all this logic to compare two images with `count_diff_between_images` (from `doc/primitives.h`). You can see an example about how to use in the [fli format](https://github.com/aseprite/aseprite/blob/master/src/app/file/fli_format.cpp#L118).